osu.ElasticIndexer/IndexQueueProcessor.cs

```diff
@@ -38,41 +38,44 @@ internal IndexQueueProcessor(string index, Client client, Action stopCallback)
 
         protected override void ProcessResults(IEnumerable<ScoreItem> items)
         {
-            var add = new List<SoloScore>();
-            var remove = new List<SoloScore>();
+            var buffer = new ProcessableItemsBuffer();
 
+            // Figure out what to do with the queue item.
             foreach (var item in items)
             {
-                if (item.Score.ShouldIndex)
-                    add.Add(item.Score);
+                if (item.ScoreId != null)
+                {
+                    buffer.ScoreIdsForLookup.Add(item.ScoreId.Value);
+                }
+                else if (item.Score != null)
+                {
+                    if (item.Score.ShouldIndex)
+                        buffer.Additions.Add(item.Score);
+                    else
+                        buffer.Deletions.Add(item.Score.id);
+                }
                 else
-                    remove.Add(item.Score);
+                {
+                    Console.WriteLine(ConsoleColor.Red, "queue item missing both data and action");
+                }
             }
 
-            var bulkDescriptor = new BulkDescriptor()
-                                 .Index(index)
-                                 .IndexMany(add)
-                                 .DeleteMany(remove);
+            // Handle any scores that need a lookup from the database.
+            performDatabaseLookup(buffer);
 
-            var response = client.ElasticClient.Bulk(bulkDescriptor);
+            Debug.Assert(buffer.ScoreIdsForLookup.Count == 0);
```

Comment:
seems like a strange assertion when the last line in `performDatabaseLookup` is `buffer.ScoreIdsForLookup.Clear()`